### DISCUSSION TITLE
Let api route functions return the data

### DISCUSSION THREAD

**Discussion by timsuchanek** [vercel/next.js#15746] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

When using `getStaticProps` or `getServerSideProps`, I can return data in that function:
```ts
export async function getStaticProps({ params }) {
  const user = await prisma.user.findOne({
    where: {
      id: Number(params.id),
    }
  })
  return {
    props: {
      user,
    },
  }
}
```
That is very useful, because now I can infer the return types of that function for my props.
However, from API routes I don't get the types.

## Describe the solution you'd like

I would like to be able to return a Promise in an API route:
```ts
const { PrismaClient } = require('@prisma/client')

const client = new PrismaClient()
export default async (req, res) => {
  const data = await client.user.findMany()
  return data
}
```
Now I can import the type in my component and use it.

```ts
import UsersApi from '../api/users'

export default function Users({ users }: InferTypes<typeof UsersApi>) {
  return <div>...</div>
}
```

## Describe alternatives you've considered

I'm not aware of an alternative for this.

**Comment by mAAdhaTTah** [comment]
I think you could just use `ReturnType` here, plus something like this type to unwrap the Promise:

```typescript
type PromiseValue<T> = T extends PromiseLike<infer U> ? U : never
```

There isn't a need for Next to do this for you.

**Comment by timsuchanek** [comment]
@mAAdhaTTah Yes, I'm aware of the `ReturnType` utility type.
However, that type is only usable, if you have a function, that actually returns a value.
I probably didn't make it clear in the issue, but this is how you write an API route right now:

```ts
const default async (req, res) => {
  res.json({ hello: 'world' })
}
```

For this you can't leverage the return type as the function doesn't return anything.

You could for sure do the following:

```ts
export type ReturnValue = {
  hello: string
}

const default async (req, res) => {
  res.json<ReturnValue>({ hello: 'world' })
}
```
However, then I need to manually type the response, which is not what I want.

**Comment by mrclay** [comment]
In your API, use the type `NextApiHandler` with a parameter of the type you want to receive at the client. That at least allows type-checking your API route.

But you're really asking for some kind of sugar to turn API routes into async function modules. I bet there's already a solution to do this.
